python/ray/autoscaler/_private/node_provider_availability_tracker.py

```diff
@@ -0,0 +1,201 @@
+        """Implement marshalling from nested fields. pydantic isn't a core dependency
+        so we're implementing this by hand instead."""
+        parsed = {}
+
+        node_availabilites_dict = fields.get("node_availabilities", {})
+
+        for node_type, node_availability_record_dict in node_availabilites_dict.items():
+            unavailable_information_dict = node_availability_record_dict.pop(
+                "unavailable_node_information", None
+            )
+            unavaiable_information = None
+            if unavailable_information_dict is not None:
+                unavaiable_information = UnavailableNodeInformation(
+                    **unavailable_information_dict
+                )
+
+            parsed[node_type] = NodeAvailabilityRecord(
+                unavailable_node_information=unavaiable_information,
+                **node_availability_record_dict,
+            )
+
+        return NodeAvailabilitySummary(node_availabilities=parsed)
+
+    def summary_string(self, separator_len: int) -> str:
+        if self:
+            formatted_lines = []
+            sorted_keys = sorted(
+                self.node_availabilities.keys(),
+                key=lambda node_type: self.node_availabilities[
+                    node_type
+                ].last_checked_timestamp,
+            )
+            for node_type in sorted_keys:
+                record = self.node_availabilities[node_type]
+                category = "Available"
+                if not record.is_available:
+                    assert record.unavailable_node_information is not None
+                    category = record.unavailable_node_information.category
+                attempted_time = datetime.datetime.fromtimestamp(
+                    record.last_checked_timestamp
+                )
+                formatted_time = (
+                    f"{attempted_time.hour}:"
+                    f"{attempted_time.minute}:"
+                    f"{attempted_time.second}"
+                )
+                formatted_line = (
+                    f" {node_type} (attempted={formatted_time}): {category}"
+                )
+                formatted_lines.append(formatted_line)
+
+            return (
+                "Launches\n"
+                + ("-" * separator_len)
+                + "\n"
+                + "Node types:\n"
+                + ("\n".join(formatted_lines))
+            )
+
+        else:
+            return ""
+        pass
+
+    def __eq__(self, other: "NodeAvailabilitySummary"):
+        return self.node_availabilities == other.node_availabilities
+
+    def __bool__(self) -> bool:
+        return bool(self.node_availabilities)
+
+
+class NodeProviderAvailabilityTracker:
+    """A thread safe, TTL cache of node provider availability. We don't use
+    cachetools.TTLCache because we want fine grain control over when entries
+    expire (e.g. insert an entry at a previous point in time)."""
```

Comment:
I don't 100% follow the why-not-ttlcache explanation.
(Not saying TTLCache is the way to go, just not parsing the explanation.)

Reply:
I've updated the description a bit, let me know if it makes more sense?

Reply:
yup